docs: contributing guide (#1605)
Arize-ai/phoenix: CONTRIBUTING.md

-   Follow the [development guide](./DEVELOPMENT.md) to setup your local environment.
-   If you've fixed a bug or added code that should be tested, add tests!
-   Ensure test suite pass. (`hatch run tests` and `npm run test` for app changes)
-   Make sure your code is formatted with `hatch run style:fix` and `npm run prettier` for app changes.
-   Make sure to your code lints with `npm run lint` for app changes.
-   Run type checking with `hatch run type:check` and `npm run typecheck` for app changes.

### Pull Request (PR) Descriptions

A PR description is a public record of what change is being made and why it was made. It will become a permanent part of our version control history, and will possibly be read by many people other than your reviewers over the years. Follow the following guidelines when writing a PR description:

-   Title: The tile must conform to [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) format and must sufficiently describe the change. Since PR titles are used to form release notes, titles with generic or non-descriptive content (“Fix build.”, “Add patch.”) are not allowed.

> -   Issue: The first line of the description should contain a reference to the issue that the PR is solving. For example, `fixes #1234` or `resolves #1234`. While this is not required for urgent fixes, it is required for all other PRs so that the issue is clearly tracked and triaged by a core team member.

Reviewer: Does this need to be part of the PR description itself? I usually link the issue in the `Development` section, but don't leave a link in the PR description.

Author: yeah it's the easiest way to do it and will probably be added to the PR template long term.

-   Simpler to roll back

Note that reviewers have discretion to reject your change outright for the sole reason of it being too large. Usually they will thank you for your contribution but request that you somehow make it into a series of smaller changes. It can be a lot of work to split up a change after you've already written it, or require lots of time arguing about why the reviewer should accept your large change. It's easier to just write small PRs in the first place.

For a comprehensive guide on how to write small PRs, see this [guide](https://github.com/google/eng-practices/blob/master/review/developer/small-cls.md)

## Contributor License Agreement (CLA)

In order to accept your pull request, we need you to submit a CLA. You only need to do this once. Simply reply to your first pull request with `I have read the CLA Document and I hereby sign the CLA`. Your signature will be recorded automatically in the `cla` branch.

## Code Reviews

A code review is a process where someone other than the author(s) of a piece of code examines that code. Code committed to the codebase is both the responsibility of the author and the reviewer. Code reviews should look at:

> 

Reviewer 2: It might be helpful to clarify the meaning of an approving review? I think there's many good ways to spell this sort of thing out; here's a rough example of one I liked in the past: "A review approval means both that the content of the PR is relevant and improves `phoenix` and that reviewer feels the core contributing team would be comfortable and capable of maintaining this code if we discover a bug or it needs to change."

Author: I'm also hesitant to spell this out as leaving room for some level of subtlety is okay here - where if the quality of reviews goes up because of the core aspects of a review spelled out as below are covered - the confidence in the maintainability / usefulness of said code will be an after thought. I also don't want a a bar per-se as it will ultimately be used as a defensive moat rather than a catalyst. I think we collectively agree with the above sentiment you spell out and am confident we don't need guidance there. I'm going to leave this out for now as I think if we can properly adhere to a simple set of what to look for and some basic guidelines on how to conduct ourselves, things will fall into place.

Reviewer 2: makes sense to me!

-   Design: Is the code well-designed and appropriate for your system?
-   Functionality: Does the code behave as the author likely intended? Is the way the code behaves good for its users?
-   Complexity: Could the code be made simpler? Would another developer be able to easily understand and use this code when they come across it in the future? Can multiple
-   Tests: Does the code have correct and well-designed automated tests?
-   Naming: Did the developer choose clear names for variables, classes, methods, etc.?
-   Comments: Are the comments clear and useful?
-   Style: Does the code follow our style guides? Note, in most cases, style nits should be avoided and be enforced entirely by automated tooling. However some stylistic decisions can be discussed if it impacts readability and complexity.
-   Documentation: Did the developer also update relevant documentation?

> All of the above are grounds for a reviewer to request changes in a PR. Consensus should be reached to the best of the abilities of the author and reviewer. However, if consensus cannot be reached between the two parties, the review should be escalated to the technical lead.

Reviewer 3: ```suggestion
All of the above are grounds for a reviewer to request changes in a PR. Consensus should be reached to the best of the abilities of the author(s) and reviewer. However, if consensus cannot be reached between the two parties, the review should be escalated to the technical lead.
```

> All of the above are grounds for a reviewer to request changes in a PR. Consensus should be reached to the best of the abilities of the author(s) and reviewer. However, if consensus cannot be reached between the two parties, the review should be escalated to the technical lead.

Reviewer 3: [source](https://google.github.io/eng-practices/review/reviewer/standard.html#:~:text=it%20is%20the,accomplish%20their%20goals.)
```suggestion
All of the above are grounds for a reviewer to request changes in a PR. It is the duty of the reviewer to make sure that each PR is of such a quality that the overall code health of the codebase is not decreasing as time goes on. This can be tricky, because often, codebases degrade through _small_ decreases in code health over time. Consensus should be reached to the best of the abilities of the author(s) and reviewer. However, if consensus cannot be reached between the two parties, the review should be escalated to the technical lead.
```

Author: Similar to the above, while I don't disagree, I don't think it needs spelling out at the current moment and don't think the addition would necessarily change the way we operate.

### Code Review Conduct

-   Be kind.
-   Explain your reasoning.
-   Balance giving explicit directions with just pointing out problems and letting the developer decide.
-   Encourage developers to simplify code or add code comments instead of just explaining the complexity to you.

> Remember that you are both on the same team and are working towards the same goal. The goal is not to make the code perfect, but to make it better. Make sure your comments are constructive and actionable. Treat the code diff as an non-personal artifact under inspection rather than a personal creation. For a comprehensive guide on how to provide effective PR feedback, see this [guide](https://google.github.io/eng-practices/review/reviewer/comments.html).

Reviewer 3: ```suggestion
Remember that you are both on the same team and are working towards the same goal. The goal is not to make the code perfect, but to make it better. Make sure your review comments are constructive and actionable. Treat the code contribution as a communal artifact under construction rather than a personal creation under evaluation. For a comprehensive guide on how to provide effective PR feedback, see this [guide](https://google.github.io/eng-practices/review/reviewer/comments.html).
```

> Remember that you are both on the same team and are working towards the same goal. The goal is not to make the code perfect, but to make it better. Make sure your review comments are constructive and actionable. Treat the code contribution as a communal artifact under construction rather than a personal creation under evaluation. For a comprehensive guide on how to provide effective PR feedback, see this [guide](https://google.github.io/eng-practices/review/reviewer/comments.html).

Reviewer 3: [source](https://google.github.io/eng-practices/review/reviewer/standard.html#:~:text=Aspects%20of%20software,of%20the%20system.)

[source](https://google.github.io/eng-practices/review/reviewer/standard.html#:~:text=Reviewers%20should%20always%20feel%20free%20to%20leave%20comments%20expressing%20that%20something%20could%20be%20better%2C%20but%20if%20it%E2%80%99s%20not%20very%20important%2C%20prefix%20it%20with%20something%20like%20%E2%80%9CNit%3A%20%E2%80%9C%20to%20let%20the%20author%20know%20that%20it%E2%80%99s%20just%20a%20point%20of%20polish%20that%20they%20could%20choose%20to%20ignore.)
```suggestion
Remember that you are both on the same team and are working towards the same goal. The goal is not to make the code perfect, but to make it better. Make sure your review comments are constructive and actionable. Treat the code contribution as a communal artifact under construction rather than a personal creation under evaluation. 

Aspects of software design are almost never a pure style issue or just a personal preference. They are based on underlying principles and should be weighed on those principles, not simply by personal opinion. Sometimes there are a few valid options. If the author can demonstrate (either through data or based on solid engineering principles) that several approaches are equally valid, then the reviewer should accept the preference of the author. Otherwise the choice is dictated by standard principles of software design. If no other rule applies, then the reviewer may ask the author to be consistent with what is in the current codebase, as long as that doesn’t worsen the overall code health of the system.

Reviewers should _always_ feel free to leave comments expressing that something could be better, but if it’s not very important, prefix it with something like “Nit: “ to let the author know that it’s just a point of polish that they could choose to ignore.

For a comprehensive guide on how to provide effective PR feedback, see this [guide](https://google.github.io/eng-practices/review/reviewer/comments.html).
```

Author: While I don't disagree with some of the sentiment and source material you are highlighting I don't want to spell out every nuanced aspect of PRs - largely because I don't want this to be an artifact that is doctrine. I think this is why I linked out to the guidelines in places where nuance can be inferred but am going to not add to this document as is because it's brevity is more important right now than it's completeness.  I'm gonna optemize for simplicity to start and can extend it as needed.

Reviewer 3: ok. sounds good. thanks
